crates/oxc_transformer/src/plugins/replace_global_defines.rs

```diff
@@ -220,26 +220,55 @@ pub struct ReplaceGlobalDefinesReturn {
 pub struct ReplaceGlobalDefines<'a> {
     allocator: &'a Allocator,
     config: ReplaceGlobalDefinesConfig,
+    ast_node_lock: Option<Address>,
 }
 
 impl<'a> Traverse<'a> for ReplaceGlobalDefines<'a> {
     fn enter_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
-        self.replace_identifier_defines(expr, ctx);
-        self.replace_dot_defines(expr, ctx);
+        if self.ast_node_lock.is_some() {
+            return;
+        }
+        let mut flag = self.replace_identifier_defines(expr, ctx);
+        flag |= self.replace_dot_defines(expr, ctx);
+        if flag {
+            self.ast_node_lock = Some(expr.address());
+        }
+    }
+
+    fn exit_expression(&mut self, node: &mut Expression<'a>, _ctx: &mut TraverseCtx<'a>) {
+        if self.ast_node_lock == Some(node.address()) {
+            self.ast_node_lock = None;
+        }
     }
 
     fn enter_assignment_expression(
         &mut self,
         node: &mut AssignmentExpression<'a>,
         ctx: &mut TraverseCtx<'a>,
     ) {
-        self.replace_define_with_assignment_expr(node, ctx);
+        if self.ast_node_lock.is_some() {
+            return;
+        }
+        if self.replace_define_with_assignment_expr(node, ctx) {
+            // `AssignmentExpression` is always stored in a `Box`, since it came from Traverse Visitor
```

Comment:
It is true that `AssignmentExpression` is always stored in a `Box`, but the reason for that isn't that it came from the `Traverse` visitor. The reason is that the only place `AssignmentExpression` appears in AST types is in:

```rs
enum Expression<'a> {
  // ...
  AssignmentExpression(Box<'a, AssignmentExpression<'a>>),
  // ...
}
```

This doesn't generalize to all types that come from the `Traverse` visitor. e.g. `IdentifierName` isn't always stored in a `Box` (e.g. in `StaticMemberExpression`).

Not trying to lecture you! Just explaining, as I know you're interested in using `GetAddress` in Rolldown, and it's useful to know its limitations/gotchas.

Reply:
thanks